### DISCUSSION TITLE
What is the history of the no-braces for single line statements feature?

### DISCUSSION THREAD

**Discussion by douglasg14b** [dotnet/roslyn#55121] [Q&A] [answered]
I've ran into many opinions on this, several of which seems to assume the intention of this feature "is to be used". I have been unable to find a history for this, or even a doc that names the feature explicitly, and want to better understand the intentions behind it.

This isn't really meant to be a discussion if it should or should not be used, except where that pertains to the history & intention of the feature. I believe the reasoning for & against have already been well established in several blogs & SO threads.

**Example:**

```cs

if (something)
    DoAThing()
    
if (something)
{
    DoAThing()
}
```

I also found this related linting rule: [IDE0011 - csharp_prefer_brace](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0011) which by default requires braces (Though the severity is `silent`). 

## My questions are:

1. What is the 'formal' name of this language feature
2. What is the history of it, when was this introduced, and what where the intentions with it's introduction?
3. Given there is a `silent` default rule against using it, was there some conclusion on the style that led to that rule being created?
    * Bonus Question: Why/When was the rule created, and why is the severity `silent` instead of `suggestion`?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> What is the 'formal' name of this language feature

"The if statement": https://github.com/dotnet/csharplang/blob/main/spec/statements.md#the-if-statement

> What is the history of it, when was this introduced, and what where the intentions with it's introduction?

It was introduced in the first version of the language.  The intention was to allow if-statements :)

> Why/When was the rule created, and why is the severity silent instead of suggestion?

I create the rule.  It was created early in the days or roslyn.  Likely around 5-10 years ago.  It's silent because we don't think we should push our views on people that this should be changed.    It defaults to the direction it has because when we analzyed the ecosystem, the fast majority of use cases used this style.

**Reply by douglasg14b** [reply]
> "The if statement"

Fair enough 😅 

I, for some reason, expected the alternative "brace less" syntax, and the control flow rules that follow, to be labeled separately.

Thanks for the reply, I appreciate your answer.

**Reply by CyrusNajmabadi** [reply]
> I, for some reason, expected the alternative "brace less" syntax,

Nope :)  From teh language perspective (as @HaloFour points out) it's just the if-statemnet, and that construct can just contain an embedded-statement.  Expression-statements (like `DoAThing();`) are embedded statements, as are `blocks` (`{ ... }`).  

This has a long root in the C-lineage of languages.

**Reply by douglasg14b** [reply]
Gotcha, so to make sure I understand:

This isn't so much a distinguished feature (Being able to have an embedded statement without declaring a block with braces) , as it is "just the way it was". If there was any intention behind it, it was more along the lines of supporting a syntax other C-Syntax languages use & support, that may have been considered worthwhile to have at that time.

The rule was created and the default selected based on the conventions found when looking through C# code-bases. It not being visible is to avoid enforcing a specific style (Is this tendency the norm for most style-based rules that we have by default?)

**Reply by CyrusNajmabadi** [reply]
> as it is "just the way it was". 

Yup

> If there was any intention behind it, it was more along the lines of supporting a syntax other C-Syntax languages use & support, that may have been considered worthwhile to have at that time.

Yup.  Note: we would still consider this worthwhile right now.  In other words, it's not a thing we did in the past that we've come to dislike/regret.  It still feels like teh right hting for the language and would likely be something we would maintain if we were ever to 'start over again'.

> The rule was created and the default selected based on the conventions found when looking through C# code-bases.

Yup.

> It not being visible is to avoid enforcing a specific style 

Yup.

> Is this tendency the norm for most style-based rules that we have by default?

Not exactly.  There are plenty of times when we create a new style and genuinely believe the new style is just strictly better than whatever the previous idiomatic approach was.  In those cases, we'll enable the style and put it at 'suggestion' level by default.  Examples of that would be preferring `a?.b` over `a == null ? null : a.b`.  However, for existing things *where* we see no clear consensus on what hte ecosystem prefers, we hold the position of "it's not our place to say."

**Comment by HaloFour** [comment]
I don't think it has a name.  It's a quirk that C# inherited from C syntax.  In short, various statements like `if` can be followed by an `embedded_statement`.  A `block` is one form of `embedded_statement`.  The use of braces has to do with declaring a block, not with the conditional `if` statement.

While being able to use statements other than blocks following an `if` statement has been known to result in bugs, it's very dependent on a lot of factors including structure and style of the code and a one-size-fits-all warnings or style rule wouldn't apply.  For example, I frequently do the following:

```cs
if (condition) return value;
// or
if (condition) throw new Exception;
```

As long as both statements and the condition expression fit easily onto a single line I don't see the value in adding the braces, especially following the Allman convention.  Others may disagree with me and would always include the braces.

**Reply by douglasg14b** [reply]
I do enjoy using the example syntax for state guarding where full blocks would be.... annoying to read

```cs
        internal IStatus SetName(string name)
        {
            if (string.IsNullOrWhiteSpace(name)) return Status.FromError(CompanyMessages.Errors.SetName_BlankOrNull); 
            if (!name.IsLooseAlphaNumeric(Config.permittedNameChars)) return Status.FromError(CompanyMessages.Errors.SetName_NonAscii);
            if (name.HasMultipleConsecutiveSpaces()) return Status.FromError(CompanyMessages.Errors.SetName_MultipleConsecutiveChars);
            if (name.Length > Config.MaxNameLength) return Status.FromError(CompanyMessages.Errors.SetName_TooLong);
            if (name.Length < Config.MinNameLength) return Status.FromError(CompanyMessages.Errors.SetName_TooShort);

            Name = name;

            return Status.AsSuccess();
        }
```

**Note: Not submitting this for code review**

**Reply by douglasg14b** [reply]
Is there a linting rule that is nuanced enough to allow for the literal single-line while disallowing the bracket-less two-line block?

**Reply by HaloFour** [reply]
> bracket-less two-line block

What does that mean?

**Reply by douglasg14b** [reply]
Sorry if that was ambiguous, still unsure of the verbiage to distinguish between them...

With brackets:

```cs
if (something)
{
    DoAThing()
}
```

Without Brackets

```cs
if (something)
    DoAThing()
```

**Comment by CyrusNajmabadi** [comment]
Note: i moved this to Roslyn from CSharpLang as this seems to a question on style choices and how the tooling works around that.  There are definitely some lang questions here, but it seems to be more weighted to the tools.
